manuals/bfoverlaymount.1
.Dd August  3, 2026
.Dt BFOVERLAYMOUNT 1
.Os
.Sh NAME
.Nm bfoverlaymount
.Nd mounts basic file overlays
.Sh SYNOPSIS
.Nm bfoverlaymount
.Op Fl X Ar extended_options
.Op Fl hvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm bfoverlaymount
is a utility to mount basic file overlays
.Pp
.Nm bfoverlaymount
is part of the
.Nm libbfoverlay
package.
.Nm libbfoverlay
is a library to provide basic file overlay support
.Pp
.Ar source
basic file overlays
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr, while bfoverlaymount will remain running in the \
foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# bfoverlaymount descriptor_file /mnt/fuse
bfoverlaymount 20260803
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libbfoverlay/issues
.Sh COPYRIGHT
Copyright (C) 2020-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* library
** add indication to specify all remaining data
** complete open data file IO handles function
** add support to read and cache in memory for slow media?

* COW file format
** add base block number to allocation table block header?
** use vshadow block tree approach as runtime cache of allocation table?

* mount tool
** add suffix option
** debug log calls to unimplemented fuse functions

* tests
** complete handle tests
** read and write tests

* document examples
** virtualize 0-byte filled data
** data file with specific offset to expose specific volume
** open dirty file system with write access, create "shadow" file

20201124
* see `git log' for more recent change log
* Initial version based on libbfio and libsigscan



bfoverlaytools/bfoverlaymount.c
/*
 * Mounts basic file overlays.
 *
 * Copyright (C) 2020-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "bfoverlaytools_getopt.h"
#include "bfoverlaytools_i18n.h"
#include "bfoverlaytools_libbfoverlay.h"
#include "bfoverlaytools_libcerror.h"
#include "bfoverlaytools_libclocale.h"
#include "bfoverlaytools_libcnotify.h"
#include "bfoverlaytools_output.h"
#include "bfoverlaytools_signal.h"
#include "bfoverlaytools_unused.h"
#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"

mount_handle_t *bfoverlaymount_mount_handle = NULL;
int bfoverlaymount_abort                    = 0;

/* Signal handler for bfoverlaymount
 */
void bfoverlaymount_signal_handler(
      bfoverlaytools_signal_t signal BFOVERLAYTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "bfoverlaymount_signal_handler";

	BFOVERLAYTOOLS_UNREFERENCED_PARAMETER( signal )

	bfoverlaymount_abort = 1;

	if( bfoverlaymount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     bfoverlaymount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description               = \
		"Use bfoverlaymount to mount basic file overlays.";

	bfoverlaytools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'T', "file", "track IO traces in a seperate file" },
		{ 'v', NULL, "verbose output to stderr, while bfoverlaymount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "descriptor file", "basic file overlay descriptor file" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	libbfoverlay_error_t *error                    = NULL;
	system_character_t *option_io_trace_file       = NULL;
	const system_character_t *path_prefix          = NULL;
	const system_character_t *path_suffix          = NULL;
	system_character_t *source                     = NULL;
	char *program                                  = "bfoverlaymount";
	system_integer_t option                        = 0;
	size_t path_prefix_size                        = 0;
	size_t path_suffix_size                        = 0;
	int number_of_options                          = (int) ( sizeof( options ) / sizeof( bfoverlaytools_option_t ) );
	int verbose                                    = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point                = NULL;
	int result                                     = 0;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations bfoverlaymount_fuse_operations;

	system_character_t *option_extended_options    = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                           = { program, NULL };
	struct fuse_args bfoverlaymount_fuse_arguments = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args bfoverlaymount_fuse_arguments = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *bfoverlaymount_fuse_channel  = NULL;
#endif
	struct fuse *bfoverlaymount_fuse_handle        = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS bfoverlaymount_dokan_operations;
	DOKAN_OPTIONS bfoverlaymount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "bfoverlaytools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( bfoverlaytools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	bfoverlaytools_output_version_fprint(
	 stdout,
	 program );

	if( bfoverlaytools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = bfoverlaytools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				bfoverlaytools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				bfoverlaytools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'T':
				option_io_trace_file = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				bfoverlaytools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source descriptor file.\n" );

		bfoverlaytools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind++ ];

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		bfoverlaytools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ optind ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libbfoverlay_notify_set_stream(
	 stderr,
	 NULL );
	libbfoverlay_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &bfoverlaymount_mount_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_io_trace_file != NULL )
	{
		if( mount_handle_set_io_trace_file(
		     bfoverlaymount_mount_handle,
		     option_io_trace_file,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set IO trace file.\n" );

			goto on_error;
		}
	}
#if defined( WINAPI )
	path_prefix = _SYSTEM_STRING( "\\BFOVERLAY" );
#else
	path_prefix = _SYSTEM_STRING( "/bfoverlay" );
#endif
	path_prefix_size = 1 + system_string_length(
	                        path_prefix );

	if( mount_handle_set_path_prefix(
	     bfoverlaymount_mount_handle,
	     path_prefix,
	     path_prefix_size,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set path prefix.\n" );

		goto on_error;
	}
	if( path_suffix != NULL )
	{
		path_suffix_size = 1 + system_string_length(
		                        path_suffix );

		if( mount_handle_set_path_suffix(
		     bfoverlaymount_mount_handle,
		     path_suffix,
		     path_suffix_size,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set path suffix.\n" );

			goto on_error;
		}
	}
	if( mount_handle_open(
	     bfoverlaymount_mount_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source descriptor file\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		bfoverlaymount_fuse_arguments.argc = 0;
		bfoverlaymount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &bfoverlaymount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &bfoverlaymount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &bfoverlaymount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &bfoverlaymount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	bfoverlaymount_fuse_operations.truncate   = &mount_fuse_truncate;
	bfoverlaymount_fuse_operations.open       = &mount_fuse_open;
	bfoverlaymount_fuse_operations.read       = &mount_fuse_read;
	bfoverlaymount_fuse_operations.write      = &mount_fuse_write;
	bfoverlaymount_fuse_operations.release    = &mount_fuse_release;
	bfoverlaymount_fuse_operations.opendir    = &mount_fuse_opendir;
	bfoverlaymount_fuse_operations.readdir    = &mount_fuse_readdir;
	bfoverlaymount_fuse_operations.releasedir = &mount_fuse_releasedir;
	bfoverlaymount_fuse_operations.getattr    = &mount_fuse_getattr;
	bfoverlaymount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	bfoverlaymount_fuse_handle = fuse_new(
	                              &bfoverlaymount_fuse_arguments,
	                              &bfoverlaymount_fuse_operations,
	                              sizeof( struct fuse_operations ),
	                              bfoverlaymount_mount_handle );

	if( bfoverlaymount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          bfoverlaymount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	bfoverlaymount_fuse_channel = fuse_mount(
	                               mount_point,
	                               &bfoverlaymount_fuse_arguments );

	if( bfoverlaymount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	bfoverlaymount_fuse_handle = fuse_new(
	                              bfoverlaymount_fuse_channel,
	                              &bfoverlaymount_fuse_arguments,
	                              &bfoverlaymount_fuse_operations,
	                              sizeof( struct fuse_operations ),
	                              bfoverlaymount_mount_handle );

	if( bfoverlaymount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          bfoverlaymount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_unmount(
	 bfoverlaymount_fuse_handle );

	fuse_destroy(
	 bfoverlaymount_fuse_handle );

	fuse_opt_free_args(
	 &bfoverlaymount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &bfoverlaymount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &bfoverlaymount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	bfoverlaymount_dokan_options.Version    = DOKAN_VERSION;
	bfoverlaymount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	bfoverlaymount_dokan_options.SingleThread = TRUE;
#else
	bfoverlaymount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		bfoverlaymount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		bfoverlaymount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	bfoverlaymount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	bfoverlaymount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	bfoverlaymount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	bfoverlaymount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	bfoverlaymount_dokan_operations.CreateDirectory      = NULL;
	bfoverlaymount_dokan_operations.Cleanup              = NULL;
	bfoverlaymount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	bfoverlaymount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	bfoverlaymount_dokan_operations.WriteFile            = &mount_dokan_WriteFile;
	bfoverlaymount_dokan_operations.FlushFileBuffers     = NULL;
	bfoverlaymount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	bfoverlaymount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	bfoverlaymount_dokan_operations.FindFilesWithPattern = NULL;
	bfoverlaymount_dokan_operations.SetFileAttributes    = NULL;
	bfoverlaymount_dokan_operations.SetFileTime          = NULL;
	bfoverlaymount_dokan_operations.DeleteFile           = NULL;
	bfoverlaymount_dokan_operations.DeleteDirectory      = NULL;
	bfoverlaymount_dokan_operations.MoveFile             = NULL;
	bfoverlaymount_dokan_operations.SetEndOfFile         = &mount_dokan_SetEndOfFile;
	bfoverlaymount_dokan_operations.SetAllocationSize    = NULL;
	bfoverlaymount_dokan_operations.LockFile             = NULL;
	bfoverlaymount_dokan_operations.UnlockFile           = NULL;
	bfoverlaymount_dokan_operations.GetFileSecurity      = NULL;
	bfoverlaymount_dokan_operations.SetFileSecurity      = NULL;
	bfoverlaymount_dokan_operations.GetDiskFreeSpace     = NULL;
	bfoverlaymount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	bfoverlaymount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	bfoverlaymount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	bfoverlaymount_dokan_operations.Cleanup              = NULL;
	bfoverlaymount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	bfoverlaymount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	bfoverlaymount_dokan_operations.WriteFile            = &mount_dokan_WriteFile;
	bfoverlaymount_dokan_operations.FlushFileBuffers     = NULL;
	bfoverlaymount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	bfoverlaymount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	bfoverlaymount_dokan_operations.FindFilesWithPattern = NULL;
	bfoverlaymount_dokan_operations.SetFileAttributes    = NULL;
	bfoverlaymount_dokan_operations.SetFileTime          = NULL;
	bfoverlaymount_dokan_operations.DeleteFile           = NULL;
	bfoverlaymount_dokan_operations.DeleteDirectory      = NULL;
	bfoverlaymount_dokan_operations.MoveFile             = NULL;
	bfoverlaymount_dokan_operations.SetEndOfFile         = &mount_dokan_SetEndOfFile;
	bfoverlaymount_dokan_operations.SetAllocationSize    = NULL;
	bfoverlaymount_dokan_operations.LockFile             = NULL;
	bfoverlaymount_dokan_operations.UnlockFile           = NULL;
	bfoverlaymount_dokan_operations.GetFileSecurity      = NULL;
	bfoverlaymount_dokan_operations.SetFileSecurity      = NULL;
	bfoverlaymount_dokan_operations.GetDiskFreeSpace     = NULL;
	bfoverlaymount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	bfoverlaymount_dokan_operations.Unmounted            = NULL;
	bfoverlaymount_dokan_operations.FindStreams          = NULL;
	bfoverlaymount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &bfoverlaymount_dokan_options,
	          &bfoverlaymount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount basic file overlays format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( bfoverlaymount_fuse_handle != NULL )
	{
		fuse_destroy(
		 bfoverlaymount_fuse_handle );
	}
	fuse_opt_free_args(
	 &bfoverlaymount_fuse_arguments );
#endif
	if( bfoverlaymount_mount_handle != NULL )
	{
		mount_handle_free(
		 &bfoverlaymount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

